### DISCUSSION TITLE
How to automate NanoFramework project modification: download, resource file change, NuGet addition, build, and obtain binaries in .NET Core APP?

### DISCUSSION THREAD

**Discussion by ijaznoufal** [nanoframework/Home#1403] [Q&A]
0

I'm working on a project where I need to automate the process of modifying a NanoFramework project. I want to create a generic .NET Core solution that can perform the following tasks:

1. Download a base NanoFramework project from a repository (Git or similar).
2. Modify the project's resource files, especially the client certificate, based on the client's specific needs.
3. Add NuGet packages to the project, as different clients may require different packages.
4. Build the project and obtain the binaries (binaries are typically firmware images).
5. Ideally, I'd like to parameterize this process, so I can repeat it for different clients easily.

Example :

I have a base NanoFramework firmware code in a repo with a Resource.resx file containing a client certificate. I need to automate the process of building custom firmware for different devices, each requiring a different client certificate and specific NuGet packages.

For example, I have two devices: ESP32-1 and ESP32-2. ESP32-1 needs a firmware bin with a client certificate change and specific NuGet packages for an ADXL sensor, while ESP32-2 requires a different certificate and NuGet packages for a BMP280 sensor.

How can I create a generic .NET Core solution that, on the fly, modifies the client certificate, adds the necessary NuGet packages, and builds the NanoFramework project, returning the customized firmware bin for each device?

![image](https://github.com/nanoframework/Home/assets/20909192/dbc09974-6531-4751-ac14-1ffadaea0335)
![image](https://github.com/nanoframework/Home/assets/20909192/b5004aaf-c032-4683-a453-b5474dbc48ea)

**Comment by josesimoes** [comment]
You're description above is very clear. I don't see anything preventing you from doing it. It's a matter of coding those tasks of manipulating the required project files. The build can be accomplished by a call to msbuild.

**Reply by ijaznoufal** [reply]
No just wanted to know, Changing the resource file values , adding nuget and building the project to get the bin(firmware) is it possible on whole. - On fly ??

**Reply by networkfusion** [reply]
I see these as "program" rather than "firmware" issues. You can easily use "ifdef's" to perform that in a GH workflow. This could use selective resource files or, depending on needs, could just be a file in the SRC.

For instance, I personally (at the moment) use the SD card for storage of the certs and include packages differently depending on device (see: https://github.com/networkfusion/OrgPal-Three-Demo), but, with the right GH workflow, you could just have a static class that uses defines to include the certs, nuget packages et.al.

**Reply by networkfusion** [reply]
With regards to different packages for different devices, your best bet is either creating a shared project for common code with seperate projects for each device type, or adding build definitions for nugets within the project (some of the code in the sample repos uses one of these 2 methods). Or even adding the libs used as sub-modules and call as needed.
